src/messages/MClientSession.h

```diff
@@ -31,6 +32,7 @@ class MClientSession final : public SafeMessage {
   std::map<std::string, std::string> metadata;
   feature_bitset_t supported_features;
   metric_spec_t metric_spec;
+  MDSAuthCaps auth_caps;
```

Comment:
Writing down a discussion I had with Ramana today:

I do not believe this is the right approach. While it appears to be future-proof because now the Client "knows" its auth caps and can be relied on to implement new authorizations (like root_squash), it has serious issues:

- Old clients will not understand new `MDSAuthCaps` versions. So a new client feature flag will be required for every change which requires new authorization handling in the Client.
- Clients have to know how to parse these caps. This is trivial for ceph-fuse but not for the kernel client. Parsing grammars in the kernel is something we should avoid.

Because we still need feature bits for any new authorization types, I think this should be handled as new fields (gid/uid/root_squash) in the `MClientSession` message with corresponding client feature flags. The server tells the client about these authorization restrictions. It's up to the clients to implement them. If the operator wants to enforce them, they can set the required client features [1].

[1] https://docs.ceph.com/en/latest/cephfs/administration/#required-client-features

Reply:
thanks, @batrick .  This discussion is more appropriate for the PR https://github.com/ceph/ceph/pull/48027 . I'm going to link it that there. This PR is based on that one, and adds changes to the MDS to disallow old clients that lack the feature to parse MDS auth caps.

Reply:
> Writing down a discussion I had with Ramana today:
> 
> I do not believe this is the right approach. While it appears to be future-proof because now the Client "knows" its auth caps and can be relied on to implement new authorizations (like root_squash), it has serious issues:
> 
> * Old clients will not understand new `MDSAuthCaps` versions. So a new client feature flag will be required for every change which requires new authorization handling in the Client.
> * Clients have to know how to parse these caps. This is trivial for ceph-fuse but not for the kernel client. Parsing grammars in the kernel is something we should avoid.

Yeah, for kclient it's true.

> 
> Because we still need feature bits for any new authorization types, I think this should be handled as new fields (gid/uid/root_squash) in the `MClientSession` message with corresponding client feature flags. The server tells the client about these authorization restrictions. It's up to the clients to implement them. If the operator wants to enforce them, they can set the required client features [1].

Let me check how to do with this. 
Thanks @batrick 

> 
> [1] https://docs.ceph.com/en/latest/cephfs/administration/#required-client-features

Reply:
> Because we still need feature bits for any new authorization types, I think this should be handled as new fields (gid/uid/root_squash) in the `MClientSession` message with corresponding client feature flags. The server tells the client about these authorization restrictions. It's up to the clients to implement them. If the operator wants to enforce them, they can set the required client features [1].

I think for something security-focused like this, we can't rely on the operator setting required flags. What we can do is have the MDS refuse to mount a client which doesn't understand cephx caps that involve client-side enforcement. Does that work?

Reply:
> Because we still need feature bits for any new authorization types, I think this should be handled as new fields (gid/uid/root_squash) in the `MClientSession` message with corresponding client feature flags. The server tells the client about these authorization restrictions. It's up to the clients to implement them. If the operator wants to enforce them, they can set the required client features [1].

@batrick, I remember we discussed a bit about the above approach a while back and recall that @lxbsz was intending to use MDSAuthCaps to fix a set of async dirops bugs in the kclient. Are those still valid @lxbsz? FWIW, The alternate approach to this was discussed here - https://github.com/ceph/ceph/pull/48027#issuecomment-1248151204.

Reply:
> I think for something security-focused like this, we can't rely on the operator setting required flags. What we can do is have the MDS refuse to mount a client which doesn't understand cephx caps that involve client-side enforcement. Does that work?

We can make it the default for the mons to automatically add these bits to the required client features bitset but that would probably be very surprising to have all existing client mounts get evicted on upgrade.

Reply:
> > Because we still need feature bits for any new authorization types, I think this should be handled as new fields (gid/uid/root_squash) in the `MClientSession` message with corresponding client feature flags. The server tells the client about these authorization restrictions. It's up to the clients to implement them. If the operator wants to enforce them, they can set the required client features [1].
> 
> @batrick, I remember we discussed a bit about the above approach a while back and recall that @lxbsz was intending to use MDSAuthCaps to fix a set of async dirops bugs in the kclient. Are those still valid @lxbsz? FWIW, The alternate approach to this was discussed here - [#48027 (comment)](https://github.com/ceph/ceph/pull/48027#issuecomment-1248151204).

@vshankar Yeah, it should be enough for async case. Currently I think the cephx caps are enough for async. This really could reduce a lot of coding work in kclient.

Reply:
Thanks @lxbsz. I'll have a look at the changes this week (trying to clear merge backlog right now).